### PR TITLE
 [AIRFLOW-3511] Create GCP Memorystore Redis Hook

### PR DESCRIPTION
**Pull request by ryanyuan**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow-3511](https://issues.apache.org/jira/browse/AIRFLOW-3511) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
Add a hook to connect to GCP Cloud Memorystore and perform operations such as instance creation, instance deletion, etc.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
tests.contrib.hooks.test_gcp_memorystore_hook

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Comment by ashb** [comment]
I'm not too familar with GCP, so this may not be a sensible question.

What is the use case here that you want to create and delete these resources from Airflow? My immediate thought is that this sounds like slightly the wrong fit for Airflow and a tool like Terraform would be better suited to this task

File: airflow/contrib/hooks/gcp_memorystore_hook.py [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+class MemorystoreHook(GoogleCloudBaseHook):
+    """
+    Interact with Google Cloud Memorystore. This hook uses the Google Cloud Platform
+    connection.
+
+    This object is not threads safe. If you want to make multiple requests
+    simultaneously, you will need to create a hook per thread.
+    """
+
+    def __init__(
```
Comment (mik-laj) [inline airflow/contrib/hooks/gcp_memorystore_hook.py:40]: Please add `@apply_defaults` decorator.

File: docs/integration.rst [hunk 2]
```diff
@@ -927,6 +927,16 @@ DatastoreHook
     :members:
 
 
+Cloud Memorystore
+'''''''''''''''
+
+MemorystoreHook
+"""""""""""""
+
+.. autoclass:: airflow.contrib.hooks.gcp_memorystore_hook.MemorystoreHook
```
Comment (mik-laj) [inline docs/integration.rst:936]: The structure of `integration.rst` file has changed. Can you update your change?

**Comment by mik-laj** [comment]
Hello

Do you plan to continue working on this change? I am organising tickets in JIRA related to GCP and I would like to know the current status of this change. 

This functionality seems to be interesting. The community would have been happy if you had completed this change. Can I help you in this?

Regards,

**Comment by ryanyuan** [comment]
I planned to rewrite this using the python library instead of using the discovery API. Do you want to take it since you are also going to do the operators for it? @mik-laj

**Comment by mik-laj** [comment]
It sounds good. I mark it as a my task to do.

**Pull request closed without merging**